open_rarity/rarity_ranker.py

```diff
@@ -20,35 +27,75 @@ def rank_tokens(*, token_id_to_scores: dict[str, float]) -> dict[str, int]:
-            provided tokens ids to the rank it has within the set of possible
-            tokens defined by @token_id_to_scores.
+        list[TokenRarity]
+            list of TokenRarity objects with score, rank and token information
+        """
+
+        if (
+            collection is None
+            or collection.tokens is None
+            or len(collection.tokens) == 0
+        ):
+            return collection
+
+        tokens = collection.tokens
+        scores: list[float] = scorer.score_tokens(collection, tokens=tokens)
+
+        # fail ranking if dimension of scores doesn't match dimension of tokens
+        assert len(tokens) == len(scores)
+
+        token_rarity_list: list[TokenRarity] = []
+        # augment collection tokes with score information
+        for idx, token in enumerate(tokens):
+            token_rarity_list.append(
+                TokenRarity(token=token, score=scores[idx])
+            )
+
+        return RarityRanker.rank_tokens(token_rarity_list)
+
+    @staticmethod
+    def rank_tokens(tokens: list[TokenRarity]) -> list[TokenRarity]:
+        """Ranks a set of tokens Scores that are higher indicate a higher rarity,
+        and thus a lower rank.
+        Tokens with the same score will be assigned the same rank, e.g. we use RANK
+        (vs. DENSE_RANK).
+        Example: 1, 2, 2, 2, 5.
+        Scores are considered the same rank if they are within about 9 decimal digits
+        of each other.
+
+        Parameters
+        ----------
+        tokens : list[TokenRarity]
+            unordered list of tokens with rarity score information
+
+        Returns
+        -------
+        tokens: list[TokenRarity]
+            list of tokens with rarity scores sorted by rank
+
         """
-        sorted_token_ids = sorted(
-            token_id_to_scores.keys(),
-            key=lambda k: token_id_to_scores[k],
+        sorted_tokens_by_score: list[TokenRarity] = sorted(
+            tokens,
+            key=lambda k: k.score,
             reverse=True,
         )
-        token_id_to_rank = {}
 
-        for i, token_id in enumerate(sorted_token_ids):
+        # perform ranking of each token in collection
+        for i, token in enumerate(sorted_tokens_by_score):
             rank = i + 1
             if i > 0:
-                prev_token_id = sorted_token_ids[i - 1]
-                scores_equal = math.isclose(
-                    token_id_to_scores[token_id],
-                    token_id_to_scores[prev_token_id],
-                )
+                prev_token = sorted_tokens_by_score[i - 1]
+                scores_equal = math.isclose(token.score, prev_token.score)
                 if scores_equal:
-                    rank = token_id_to_rank[prev_token_id]
+                    rank = prev_token.rank
 
-            token_id_to_rank[token_id] = rank
+            token.rank = rank
 
-        return token_id_to_rank
+        return sorted_tokens_by_score
```

Comment:
I believe this may apply elsewhere as well but have we considered ranking the tokens with something like `scipy.stats.rankdata`?

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rankdata.html

Reply:
I want to minimize dependency graph here , this is the time i would say reinventing the bicycle is needed :)